File: quad/cython_c/quad_openmp.py
from quad_openmp_wrapper import quadrature

a = 0.0
b = 10.0
exact = 0.49936338107645674464
n=1000000000

print()
print("QUAD_OPENMP:")
print("  C version")
print("  Use OpenMP for parallel execution.")
print("  Estimate the integral of f(x) from A to B.")
print("  f(x) = 50 / ( pi * ( 2500 * x * x + 1 ) ).")
print()
print("  A        = {}".format(a))
print("  B        = {}".format(b))
print("  N        = {}".format(n))
print("  Exact    = {}".format(exact))
#print("  Exact    = {}%24.16f", exact)

total, error, wtime = quadrature(a, b, exact, n)

print()
#print("  Estimate = %24.16f", total)
print("  Estimate = {}".format(total))
print("  Error    = {}".format(error))
print("  Time     = {}".format(wtime))

print()
print("QUAD_OPENMP:")
print("  Normal end of execution.")
print()

File: quad/cython_c/quad_openmp.c
# include <math.h>
# include <omp.h>

// Not all compilers have M_PI and this code is GPL,
// so I have taken the definition from GCC 4.9.
#ifndef M_PI
#define M_PI        3.14159265358979323846264338327950288
#endif

double f ( double x );

int quad (double* total, double* error, double* wtime, double a,
          double b, double exact, int n) {
  int i;
  double t;
  double x;

  *wtime = omp_get_wtime ( );

  t = 0.0;

# pragma omp parallel shared ( a, b, n ) private ( i, x ) 

# pragma omp for reduction ( + : t )

  for ( i = 0; i < n; i++ ) {
    x = ( ( double ) ( n - i - 1 ) * a + ( double ) ( i ) * b ) / ( double ) ( n - 1 );
    t = t + f ( x );
  }

  *wtime = omp_get_wtime ( ) - *wtime;
  *total = ( b - a ) * t / ( double ) n;
  *error = fabs ( *total - exact );

  return 0;
}

double f ( double x ) {
  return 50.0 / ( M_PI * ( 2500.0 * x * x + 1.0 ) );
}

